### PR TITLE
Convert `proc_macro_back_compat` lint to an unconditional error.

### PR DESCRIPTION
**Pull request by nnethercote**
We still check for the `rental`/`allsorts-rental` crates. But now if they are detected we just emit a fatal error, instead of emitting a warning and providing alternative behaviour.
    
The original "hack" implementing alternative behaviour was added in #73345.
    
The lint was added in #83127.
    
The tracking issue is #83125.
    
The direct motivation for the change is that providing the alternative behaviour is interfering with #125174 and follow-on work.

r? @estebank

### CODE REVIEW

**Comment by nnethercote** [comment]
I'm not 100% certain this is the right way to do this, but it seems like a reasonable first attempt.

cc @petrochenkov @Aaron1011

**Comment by estebank** [comment]
@bors try

Gonna do a crater run as a sanity check that we haven't missed something, but otherwise I'm ok with landing this change as is. We provide actionable user feedback while removing the hack and without blocking your other work. This is the best possible resolution, beyond a more advanced/generalized mechanism to handle popular dependency versions with known issues.

**Review by estebank — APPROVED** [review]

**Comment by nnethercote** [comment]
I rebased.

@bors try

**Comment by nnethercote** [comment]
@craterbot check

File: compiler/rustc_expand/src/base.rs [hunk 1]
```diff
@@ -1330,80 +1329,63 @@ pub fn parse_macro_name_and_helper_attrs(
...
+        } else {
+            let mut version = c.trim_start_matches("rental-").split('.');
+            version.next() == Some("0")
+                && version.next() == Some("5")
+                && version.next().and_then(|c| c.parse::<u32>().ok()).is_some_and(|v| v < 6)
+        };
+
+        if crate_matches {
+            // FIXME: make this translatable
+            #[allow(rustc::untranslatable_diagnostic)]
```
Comment (Nilstrieb) [inline compiler/rustc_expand/src/base.rs:1360]: this looks very translatable now

**Comment by nnethercote** [comment]
Crater results are very much as expected.

The vast majority of the failures have the expected message:
> error: using an old version of `rental`
> |
> = note: older versions of the `rental` crate no longer compile; please update to `rental` v0.5.6, or switch to one of the `rental` alternative

A few had "no space left on device" (just running out of disk space):
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/knirb.rust-tutorial/log.txt
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/robinlimit.rust_learning_basic_level/log.txt
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/to-json.testchamber/log.txt
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/yoelbassin.juliText/log.txt
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/robinlimit.rust_learning_basic_level/log.txt

One had "error: failed to download `aho-corasick v1.1.3`" (clearly a network error)
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/reg/moon_pdk_api-0.0.6/log.txt

Two had "collect2: fatal error: ld terminated with signal 7 [Bus error], core dumped" (linker crash?!)
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/zainokta.log-parser/log.txt
- https://crater-reports.s3.amazonaws.com/pr-125596/try%23c57a122e6bc7d3e243f3b2a000b6916cb51ef149/gh/znmead.guessing_game/log.txt

Other notes
- Most of the rental failures are in personal repos, presumably for long-abandoned personal projects.
- Only 11 are are in published crates. There are all at least 3 years old.

This all looks fine to me. @estebank, what do you think?

**Pull request merged**